src/integration/resources/inprocess/coordinator.go

```diff
@@ -0,0 +1,278 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package inprocess contains code for spinning up M3 resources in-process for
+// the sake of integration testing.
+package inprocess
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/cenkalti/backoff/v3"
+	"go.uber.org/zap"
+	"gopkg.in/yaml.v2"
+
+	"github.com/m3db/m3/src/cmd/services/m3query/config"
+	"github.com/m3db/m3/src/integration/resources"
+	"github.com/m3db/m3/src/integration/resources/common"
+	nettest "github.com/m3db/m3/src/integration/resources/net"
+	"github.com/m3db/m3/src/query/generated/proto/admin"
+	"github.com/m3db/m3/src/query/generated/proto/prompb"
+	"github.com/m3db/m3/src/query/server"
+	xconfig "github.com/m3db/m3/src/x/config"
+)
+
+type coordinator struct {
+	cfg     config.Configuration
+	client  common.CoordinatorClient
+	logger  *zap.Logger
+	tmpDirs []string
+
+	interruptCh chan<- error
+	shutdownCh  <-chan struct{}
+}
+
+// CoordinatorOptions are options for starting a coordinator server.
+type CoordinatorOptions struct {
+	// Logger is the logger to use for the coordinator. If not provided,
+	// a default one will be created.
+	Logger *zap.Logger
+}
+
+// NewCoordinatorFromConfigFile creates a new in-process coordinator based on the config file
+// and options provided.
+func NewCoordinatorFromConfigFile(pathToCfg string, opts CoordinatorOptions) (resources.Coordinator, error) {
+	var cfg config.Configuration
+	if err := xconfig.LoadFile(&cfg, pathToCfg, xconfig.Options{}); err != nil {
+		return nil, err
+	}
+
+	return NewCoordinator(cfg, opts)
+}
+
+// NewCoordinatorFromYAML creates a new in-process coordinator based on the YAML configuration string
+// and options provided.
+func NewCoordinatorFromYAML(yamlCfg string, opts CoordinatorOptions) (resources.Coordinator, error) {
+	var cfg config.Configuration
+	if err := yaml.Unmarshal([]byte(yamlCfg), &cfg); err != nil {
+		return nil, err
+	}
+
+	return NewCoordinator(cfg, opts)
+}
+
+// NewCoordinator creates a new in-process coordinator based on the configuration
+// and options provided.
+func NewCoordinator(cfg config.Configuration, opts CoordinatorOptions) (resources.Coordinator, error) {
+	// Replace any "0" ports with an open port
+	cfg, err := updateCoordinatorPorts(cfg)
```

Comment:
One convention I've settled on is that, if you provide an address with a 0 port (e.g. `<addr>:0`), we'll automatically choose a port for you. This follows a convention similar to what [net.Listen](https://pkg.go.dev/net#Listen) does. The idea here is this will make it much easier to potentially spin up multiple components with the same config (without having to muck with ports).

Reply:
Can you add this and the file path conversion stuff to a top level comment? Shouldn't need to read the source code to understand this.

Reply:
Yeah, I was gonna wait and write up a guide around this, but I'll just do it now